{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56167a00",
   "metadata": {},
   "source": [
    "# GeneBank data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6373dd8-82a9-41d6-ae67-9e52b19491b7",
   "metadata": {},
   "source": [
    "Dataset for evaluation of Taxonomy check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3b8addd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# Define the main directory and genebank_test_data directory\n",
    "main_dir = os.getcwd()\n",
    "genebank_test_data = os.path.join(main_dir, \"genebank_test_data\")\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "os.makedirs(genebank_test_data, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fb91c5bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 14470  100 14470    0     0  64656      0 --:--:-- --:--:-- --:--:-- 64887\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Downloading dfast_file_downloader.py (if not present) which is needed for downloading the genomes later on.\n",
    "if [ -f \"dfast_file_downloader.py\" ]; then\n",
    "   echo \"dfast_file_downloader.py exists, no need to download it again. skipping...\"\n",
    "else\n",
    "   curl -o dfast_file_downloader.py https://raw.githubusercontent.com/nigyta/dfast_core/master/scripts/dfast_file_downloader.py\n",
    "fi\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "178a9d0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the directory containing our genomes.\n",
    "genome_directory = \"/Users/mohamed/Desktop/dfast_qc/dqc_reference/genomes\"\n",
    "# List all genomes in the directory.\n",
    "all_genomes = os.listdir(genome_directory)\n",
    "for index, genome_file in enumerate(all_genomes):\n",
    "    # replace \".fna.gz\" with blank space.\n",
    "    all_genomes[index] = genome_file.replace(\".fna.gz\", \"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9a9f333f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eaa1820",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genebank_test_data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6 1105M    6 71.3M    0     0   182k      0  1:43:10  0:06:39  1:36:31  142k\n",
      "curl: (18) transfer closed with 1084423396 bytes remaining to read\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  593M  100  593M    0     0  8198k      0  0:01:14  0:01:14 --:--:-- 23.0M\n",
      "head: assembly_summary_genbank.txt: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "curl -o genebank_test_data/assembly_summary_genbank.txt -L https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt\n",
    "curl -o genebank_test_data/ANI_report_prokaryotes.txt -L https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/ANI_report_prokaryotes.txt\n",
    "head -5 assembly_summary_genbank.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "612c7fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_refseq = pd.read_table(\"genebank_test_data/assembly_summary_genbank.txt\", skiprows=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c476396a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# inplace is set to True, as it modifies the original DataFrame in place and does not return a new DataFrame.\n",
    "df_refseq.rename(columns={'#assembly_accession': 'assembly_accession'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "18d42134",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>replicon_count</th>\n",
       "      <th>scaffold_count</th>\n",
       "      <th>contig_count</th>\n",
       "      <th>annotation_provider</th>\n",
       "      <th>annotation_name</th>\n",
       "      <th>annotation_date</th>\n",
       "      <th>total_gene_count</th>\n",
       "      <th>protein_coding_gene_count</th>\n",
       "      <th>non_coding_gene_count</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_033763495.1</td>\n",
       "      <td>PRJNA866711</td>\n",
       "      <td>SAMN30172165</td>\n",
       "      <td>JANRKL000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2717303</td>\n",
       "      <td>2717303</td>\n",
       "      <td>Haliscomenobacter sp.</td>\n",
       "      <td>na</td>\n",
       "      <td>bins.171</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/11/10</td>\n",
       "      <td>6027</td>\n",
       "      <td>5963</td>\n",
       "      <td>51</td>\n",
       "      <td>37202810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_033763595.1</td>\n",
       "      <td>PRJNA866711</td>\n",
       "      <td>SAMN30172161</td>\n",
       "      <td>JANRKH000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2023360</td>\n",
       "      <td>2023360</td>\n",
       "      <td>Rickettsiaceae bacterium</td>\n",
       "      <td>na</td>\n",
       "      <td>bins.216</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/11/10</td>\n",
       "      <td>1245</td>\n",
       "      <td>1188</td>\n",
       "      <td>45</td>\n",
       "      <td>37202810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_033763695.1</td>\n",
       "      <td>PRJNA866711</td>\n",
       "      <td>SAMN30172155</td>\n",
       "      <td>JANRKB000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2026780</td>\n",
       "      <td>2026780</td>\n",
       "      <td>Planctomycetota bacterium</td>\n",
       "      <td>na</td>\n",
       "      <td>bins.472</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/11/10</td>\n",
       "      <td>2938</td>\n",
       "      <td>2875</td>\n",
       "      <td>53</td>\n",
       "      <td>37202810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_033763715.1</td>\n",
       "      <td>PRJNA866711</td>\n",
       "      <td>SAMN30172156</td>\n",
       "      <td>JANRKC000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>644</td>\n",
       "      <td>644</td>\n",
       "      <td>Aeromonas hydrophila</td>\n",
       "      <td>na</td>\n",
       "      <td>bins.294</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>243</td>\n",
       "      <td>243</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/11/10</td>\n",
       "      <td>4352</td>\n",
       "      <td>4206</td>\n",
       "      <td>101</td>\n",
       "      <td>37202810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_033763735.1</td>\n",
       "      <td>PRJNA866711</td>\n",
       "      <td>SAMN30172152</td>\n",
       "      <td>JANRJY000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1898104</td>\n",
       "      <td>1898104</td>\n",
       "      <td>Bacteroidota bacterium</td>\n",
       "      <td>na</td>\n",
       "      <td>bins.359</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>124</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/11/10</td>\n",
       "      <td>3550</td>\n",
       "      <td>3483</td>\n",
       "      <td>42</td>\n",
       "      <td>37202810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388866</th>\n",
       "      <td>GCA_040192195.1</td>\n",
       "      <td>PRJNA1121191</td>\n",
       "      <td>SAMN41754086</td>\n",
       "      <td>JBEFRS000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>33945</td>\n",
       "      <td>33945</td>\n",
       "      <td>Enterococcus avium</td>\n",
       "      <td>strain=YW3-43</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>105</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/06/10</td>\n",
       "      <td>4376</td>\n",
       "      <td>4272</td>\n",
       "      <td>70</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388867</th>\n",
       "      <td>GCA_040192215.1</td>\n",
       "      <td>PRJNA1121191</td>\n",
       "      <td>SAMN41754085</td>\n",
       "      <td>JBEFRR000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1353</td>\n",
       "      <td>1353</td>\n",
       "      <td>Enterococcus gallinarum</td>\n",
       "      <td>strain=YW2-39</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/06/10</td>\n",
       "      <td>3236</td>\n",
       "      <td>3104</td>\n",
       "      <td>63</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388868</th>\n",
       "      <td>GCA_040192235.1</td>\n",
       "      <td>PRJNA1121191</td>\n",
       "      <td>SAMN41754087</td>\n",
       "      <td>JBEFRT000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1353</td>\n",
       "      <td>1353</td>\n",
       "      <td>Enterococcus gallinarum</td>\n",
       "      <td>strain=JH1-3</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/06/10</td>\n",
       "      <td>3342</td>\n",
       "      <td>3195</td>\n",
       "      <td>63</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388869</th>\n",
       "      <td>GCA_040192295.1</td>\n",
       "      <td>PRJNA1121191</td>\n",
       "      <td>SAMN41754068</td>\n",
       "      <td>JBEFRA000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1354</td>\n",
       "      <td>1354</td>\n",
       "      <td>Enterococcus hirae</td>\n",
       "      <td>strain=SD2-3</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/06/10</td>\n",
       "      <td>2626</td>\n",
       "      <td>2479</td>\n",
       "      <td>68</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388870</th>\n",
       "      <td>GCA_040192315.1</td>\n",
       "      <td>PRJNA982295</td>\n",
       "      <td>SAMN39785608</td>\n",
       "      <td>JBEFVO000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>340</td>\n",
       "      <td>339</td>\n",
       "      <td>Xanthomonas campestris pv. campestris</td>\n",
       "      <td>strain=WHRI 8827A</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>130</td>\n",
       "      <td>130</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/06/10</td>\n",
       "      <td>4690</td>\n",
       "      <td>4399</td>\n",
       "      <td>145</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2388871 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        assembly_accession    bioproject     biosample         wgs_master  \\\n",
       "0          GCA_033763495.1   PRJNA866711  SAMN30172165  JANRKL000000000.1   \n",
       "1          GCA_033763595.1   PRJNA866711  SAMN30172161  JANRKH000000000.1   \n",
       "2          GCA_033763695.1   PRJNA866711  SAMN30172155  JANRKB000000000.1   \n",
       "3          GCA_033763715.1   PRJNA866711  SAMN30172156  JANRKC000000000.1   \n",
       "4          GCA_033763735.1   PRJNA866711  SAMN30172152  JANRJY000000000.1   \n",
       "...                    ...           ...           ...                ...   \n",
       "2388866    GCA_040192195.1  PRJNA1121191  SAMN41754086  JBEFRS000000000.1   \n",
       "2388867    GCA_040192215.1  PRJNA1121191  SAMN41754085  JBEFRR000000000.1   \n",
       "2388868    GCA_040192235.1  PRJNA1121191  SAMN41754087  JBEFRT000000000.1   \n",
       "2388869    GCA_040192295.1  PRJNA1121191  SAMN41754068  JBEFRA000000000.1   \n",
       "2388870    GCA_040192315.1   PRJNA982295  SAMN39785608  JBEFVO000000000.1   \n",
       "\n",
       "        refseq_category    taxid  species_taxid  \\\n",
       "0                    na  2717303        2717303   \n",
       "1                    na  2023360        2023360   \n",
       "2                    na  2026780        2026780   \n",
       "3                    na      644            644   \n",
       "4                    na  1898104        1898104   \n",
       "...                 ...      ...            ...   \n",
       "2388866              na    33945          33945   \n",
       "2388867              na     1353           1353   \n",
       "2388868              na     1353           1353   \n",
       "2388869              na     1354           1354   \n",
       "2388870              na      340            339   \n",
       "\n",
       "                                 organism_name infraspecific_name   isolate  \\\n",
       "0                        Haliscomenobacter sp.                 na  bins.171   \n",
       "1                     Rickettsiaceae bacterium                 na  bins.216   \n",
       "2                    Planctomycetota bacterium                 na  bins.472   \n",
       "3                         Aeromonas hydrophila                 na  bins.294   \n",
       "4                       Bacteroidota bacterium                 na  bins.359   \n",
       "...                                        ...                ...       ...   \n",
       "2388866                     Enterococcus avium      strain=YW3-43        na   \n",
       "2388867                Enterococcus gallinarum      strain=YW2-39        na   \n",
       "2388868                Enterococcus gallinarum       strain=JH1-3        na   \n",
       "2388869                     Enterococcus hirae       strain=SD2-3        na   \n",
       "2388870  Xanthomonas campestris pv. campestris  strain=WHRI 8827A        na   \n",
       "\n",
       "         ... replicon_count scaffold_count contig_count annotation_provider  \\\n",
       "0        ...              0            260          260                NCBI   \n",
       "1        ...              0             70           70                NCBI   \n",
       "2        ...              0             57           57                NCBI   \n",
       "3        ...              0            243          243                NCBI   \n",
       "4        ...              0            124          124                NCBI   \n",
       "...      ...            ...            ...          ...                 ...   \n",
       "2388866  ...              0            105          105                NCBI   \n",
       "2388867  ...              0             40           40                NCBI   \n",
       "2388868  ...              0             57           57                NCBI   \n",
       "2388869  ...              0             29           29                NCBI   \n",
       "2388870  ...              0            130          130                NCBI   \n",
       "\n",
       "                                           annotation_name annotation_date  \\\n",
       "0        NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/11/10   \n",
       "1        NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/11/10   \n",
       "2        NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/11/10   \n",
       "3        NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/11/10   \n",
       "4        NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/11/10   \n",
       "...                                                    ...             ...   \n",
       "2388866  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/06/10   \n",
       "2388867  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/06/10   \n",
       "2388868  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/06/10   \n",
       "2388869  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/06/10   \n",
       "2388870  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/06/10   \n",
       "\n",
       "        total_gene_count protein_coding_gene_count non_coding_gene_count  \\\n",
       "0                   6027                      5963                    51   \n",
       "1                   1245                      1188                    45   \n",
       "2                   2938                      2875                    53   \n",
       "3                   4352                      4206                   101   \n",
       "4                   3550                      3483                    42   \n",
       "...                  ...                       ...                   ...   \n",
       "2388866             4376                      4272                    70   \n",
       "2388867             3236                      3104                    63   \n",
       "2388868             3342                      3195                    63   \n",
       "2388869             2626                      2479                    68   \n",
       "2388870             4690                      4399                   145   \n",
       "\n",
       "        pubmed_id  \n",
       "0        37202810  \n",
       "1        37202810  \n",
       "2        37202810  \n",
       "3        37202810  \n",
       "4        37202810  \n",
       "...           ...  \n",
       "2388866        na  \n",
       "2388867        na  \n",
       "2388868        na  \n",
       "2388869        na  \n",
       "2388870        na  \n",
       "\n",
       "[2388871 rows x 38 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e39798bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_refseq_prok = df_refseq[df_refseq[\"group\"].isin([\"bacteria\", \"archeae\"])]\n",
    "df_refseq_prok_latest = df_refseq_prok[df_refseq_prok[\"version_status\"]==\"latest\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f8890282",
   "metadata": {},
   "outputs": [],
   "source": [
    "ani_report = pd.read_csv(\"genebank_test_data/ANI_report_prokaryotes.txt\", sep = \"\\t\")\n",
    "df_refseq_prok_latest_ok= pd.merge(df_refseq_prok_latest, ani_report[[\"taxonomy-check-status\",'# genbank-accession']], left_on='assembly_accession', right_on='# genbank-accession', how='left')\n",
    "df_refseq_prok_latest_ok = df_refseq_prok_latest_ok.drop('# genbank-accession', axis=1)\n",
    "df_refseq_prok_latest_ok = df_refseq_prok_latest_ok[df_refseq_prok_latest_ok[\"taxonomy-check-status\"]==\"OK\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6278a01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>replicon_count</th>\n",
       "      <th>scaffold_count</th>\n",
       "      <th>contig_count</th>\n",
       "      <th>annotation_provider</th>\n",
       "      <th>annotation_name</th>\n",
       "      <th>annotation_date</th>\n",
       "      <th>total_gene_count</th>\n",
       "      <th>protein_coding_gene_count</th>\n",
       "      <th>non_coding_gene_count</th>\n",
       "      <th>pubmed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_009496935.1</td>\n",
       "      <td>PRJNA579428</td>\n",
       "      <td>SAMN13111529</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>34085</td>\n",
       "      <td>34085</td>\n",
       "      <td>Riemerella anatipestifer</td>\n",
       "      <td>strain=RA-LZ01</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>State Key Laboratory of Bioreactor Engineering</td>\n",
       "      <td>Annotation submitted by State Key Laboratory o...</td>\n",
       "      <td>2019/11/03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_009496995.1</td>\n",
       "      <td>PRJNA576563</td>\n",
       "      <td>SAMN12996974</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1421</td>\n",
       "      <td>1421</td>\n",
       "      <td>Lysinibacillus sphaericus</td>\n",
       "      <td>strain=DH-B01</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2020/02/19</td>\n",
       "      <td>4544</td>\n",
       "      <td>4278</td>\n",
       "      <td>149</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_009497105.1</td>\n",
       "      <td>PRJNA577063</td>\n",
       "      <td>SAMN13018493</td>\n",
       "      <td>WIPB00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1404</td>\n",
       "      <td>1404</td>\n",
       "      <td>Priestia megaterium</td>\n",
       "      <td>strain=M4</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>241</td>\n",
       "      <td>241</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2019/10/29</td>\n",
       "      <td>5620</td>\n",
       "      <td>5396</td>\n",
       "      <td>113</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_009497195.1</td>\n",
       "      <td>PRJNA576337</td>\n",
       "      <td>SAMN12990278</td>\n",
       "      <td>WIKF00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=X21</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>137</td>\n",
       "      <td>137</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2019/10/29</td>\n",
       "      <td>5023</td>\n",
       "      <td>4754</td>\n",
       "      <td>85</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_009497315.1</td>\n",
       "      <td>PRJNA576337</td>\n",
       "      <td>SAMN12990273</td>\n",
       "      <td>WIKK00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=J22</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2019/10/29</td>\n",
       "      <td>5092</td>\n",
       "      <td>4785</td>\n",
       "      <td>83</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272758</th>\n",
       "      <td>GCA_948787075.1</td>\n",
       "      <td>PRJEB59335</td>\n",
       "      <td>SAMEA112374538</td>\n",
       "      <td>CAQFIC000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1702221</td>\n",
       "      <td>1702221</td>\n",
       "      <td>Faecalibaculum rodentium</td>\n",
       "      <td>na</td>\n",
       "      <td>MGBC100550</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>257</td>\n",
       "      <td>257</td>\n",
       "      <td>UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>Annotation submitted by UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>2023/06/28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272759</th>\n",
       "      <td>GCA_030368085.1</td>\n",
       "      <td>PRJNA962265</td>\n",
       "      <td>SAMN34393047</td>\n",
       "      <td>JASAHI000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>287</td>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>strain=R13-65</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>Sir Run Run Shaw Hospital, School of Medicine,...</td>\n",
       "      <td>Annotation submitted by Sir Run Run Shaw Hospi...</td>\n",
       "      <td>2023/06/28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272760</th>\n",
       "      <td>GCA_030368155.1</td>\n",
       "      <td>PRJNA962265</td>\n",
       "      <td>SAMN34393040</td>\n",
       "      <td>JASAHB000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>287</td>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>strain=R13-08</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>233</td>\n",
       "      <td>233</td>\n",
       "      <td>Sir Run Run Shaw Hospital, School of Medicine,...</td>\n",
       "      <td>Annotation submitted by Sir Run Run Shaw Hospi...</td>\n",
       "      <td>2023/06/28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272761</th>\n",
       "      <td>GCA_948824835.1</td>\n",
       "      <td>PRJEB59335</td>\n",
       "      <td>SAMEA112374543</td>\n",
       "      <td>CAQKXK000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>na</td>\n",
       "      <td>MGBC100555</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>549</td>\n",
       "      <td>549</td>\n",
       "      <td>UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>Annotation submitted by UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>2023/06/28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272762</th>\n",
       "      <td>GCA_948796125.1</td>\n",
       "      <td>PRJEB59335</td>\n",
       "      <td>SAMEA112374544</td>\n",
       "      <td>CAQGRD000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>297314</td>\n",
       "      <td>297314</td>\n",
       "      <td>uncultured Lachnospiraceae bacterium</td>\n",
       "      <td>na</td>\n",
       "      <td>MGBC100556</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>575</td>\n",
       "      <td>575</td>\n",
       "      <td>UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>Annotation submitted by UNIVERSITY OF CAMBRIDGE</td>\n",
       "      <td>2023/06/28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2077511 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        assembly_accession   bioproject       biosample         wgs_master  \\\n",
       "0          GCA_009496935.1  PRJNA579428    SAMN13111529                 na   \n",
       "1          GCA_009496995.1  PRJNA576563    SAMN12996974                 na   \n",
       "2          GCA_009497105.1  PRJNA577063    SAMN13018493     WIPB00000000.1   \n",
       "3          GCA_009497195.1  PRJNA576337    SAMN12990278     WIKF00000000.1   \n",
       "4          GCA_009497315.1  PRJNA576337    SAMN12990273     WIKK00000000.1   \n",
       "...                    ...          ...             ...                ...   \n",
       "2272758    GCA_948787075.1   PRJEB59335  SAMEA112374538  CAQFIC000000000.1   \n",
       "2272759    GCA_030368085.1  PRJNA962265    SAMN34393047  JASAHI000000000.1   \n",
       "2272760    GCA_030368155.1  PRJNA962265    SAMN34393040  JASAHB000000000.1   \n",
       "2272761    GCA_948824835.1   PRJEB59335  SAMEA112374543  CAQKXK000000000.1   \n",
       "2272762    GCA_948796125.1   PRJEB59335  SAMEA112374544  CAQGRD000000000.1   \n",
       "\n",
       "        refseq_category    taxid  species_taxid  \\\n",
       "0                    na    34085          34085   \n",
       "1                    na     1421           1421   \n",
       "2                    na     1404           1404   \n",
       "3                    na      562            562   \n",
       "4                    na      562            562   \n",
       "...                 ...      ...            ...   \n",
       "2272758              na  1702221        1702221   \n",
       "2272759              na      287            287   \n",
       "2272760              na      287            287   \n",
       "2272761              na      562            562   \n",
       "2272762              na   297314         297314   \n",
       "\n",
       "                                organism_name infraspecific_name     isolate  \\\n",
       "0                    Riemerella anatipestifer     strain=RA-LZ01          na   \n",
       "1                   Lysinibacillus sphaericus      strain=DH-B01          na   \n",
       "2                         Priestia megaterium          strain=M4          na   \n",
       "3                            Escherichia coli         strain=X21          na   \n",
       "4                            Escherichia coli         strain=J22          na   \n",
       "...                                       ...                ...         ...   \n",
       "2272758              Faecalibaculum rodentium                 na  MGBC100550   \n",
       "2272759                Pseudomonas aeruginosa      strain=R13-65          na   \n",
       "2272760                Pseudomonas aeruginosa      strain=R13-08          na   \n",
       "2272761                      Escherichia coli                 na  MGBC100555   \n",
       "2272762  uncultured Lachnospiraceae bacterium                 na  MGBC100556   \n",
       "\n",
       "         ... replicon_count scaffold_count contig_count  \\\n",
       "0        ...              1              1            1   \n",
       "1        ...              1              1            1   \n",
       "2        ...              0            241          241   \n",
       "3        ...              0            137          137   \n",
       "4        ...              0            139          139   \n",
       "...      ...            ...            ...          ...   \n",
       "2272758  ...              0            257          257   \n",
       "2272759  ...              0            214          214   \n",
       "2272760  ...              0            233          233   \n",
       "2272761  ...              0            549          549   \n",
       "2272762  ...              0            575          575   \n",
       "\n",
       "                                       annotation_provider  \\\n",
       "0           State Key Laboratory of Bioreactor Engineering   \n",
       "1                                                     NCBI   \n",
       "2                                                     NCBI   \n",
       "3                                                     NCBI   \n",
       "4                                                     NCBI   \n",
       "...                                                    ...   \n",
       "2272758                            UNIVERSITY OF CAMBRIDGE   \n",
       "2272759  Sir Run Run Shaw Hospital, School of Medicine,...   \n",
       "2272760  Sir Run Run Shaw Hospital, School of Medicine,...   \n",
       "2272761                            UNIVERSITY OF CAMBRIDGE   \n",
       "2272762                            UNIVERSITY OF CAMBRIDGE   \n",
       "\n",
       "                                           annotation_name annotation_date  \\\n",
       "0        Annotation submitted by State Key Laboratory o...      2019/11/03   \n",
       "1        NCBI Prokaryotic Genome Annotation Pipeline (P...      2020/02/19   \n",
       "2        NCBI Prokaryotic Genome Annotation Pipeline (P...      2019/10/29   \n",
       "3        NCBI Prokaryotic Genome Annotation Pipeline (P...      2019/10/29   \n",
       "4        NCBI Prokaryotic Genome Annotation Pipeline (P...      2019/10/29   \n",
       "...                                                    ...             ...   \n",
       "2272758    Annotation submitted by UNIVERSITY OF CAMBRIDGE      2023/06/28   \n",
       "2272759  Annotation submitted by Sir Run Run Shaw Hospi...      2023/06/28   \n",
       "2272760  Annotation submitted by Sir Run Run Shaw Hospi...      2023/06/28   \n",
       "2272761    Annotation submitted by UNIVERSITY OF CAMBRIDGE      2023/06/28   \n",
       "2272762    Annotation submitted by UNIVERSITY OF CAMBRIDGE      2023/06/28   \n",
       "\n",
       "        total_gene_count protein_coding_gene_count non_coding_gene_count  \\\n",
       "0                      0                         0                     0   \n",
       "1                   4544                      4278                   149   \n",
       "2                   5620                      5396                   113   \n",
       "3                   5023                      4754                    85   \n",
       "4                   5092                      4785                    83   \n",
       "...                  ...                       ...                   ...   \n",
       "2272758                0                         0                     0   \n",
       "2272759                0                         0                     0   \n",
       "2272760                0                         0                     0   \n",
       "2272761                0                         0                     0   \n",
       "2272762                0                         0                     0   \n",
       "\n",
       "        pubmed_id  \n",
       "0              na  \n",
       "1              na  \n",
       "2              na  \n",
       "3              na  \n",
       "4              na  \n",
       "...           ...  \n",
       "2272758        na  \n",
       "2272759        na  \n",
       "2272760        na  \n",
       "2272761        na  \n",
       "2272762        na  \n",
       "\n",
       "[2077511 rows x 38 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_refseq_prok_latest_ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b2c84f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genomes in RefSeq but not in our reference data + No type material was found. \n",
    "df_uniq_genome = df_refseq_prok_latest_ok[~df_refseq_prok_latest_ok[\"assembly_accession\"].isin(all_genomes)]\n",
    "df_uniq_genome = df_uniq_genome[df_uniq_genome[\"relation_to_type_material\"]==\"na\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "91dd1ad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Candidatus, uncultured を除く\n",
    "def is_valid_name(S):\n",
    "    organism = S[\"organism_name\"]\n",
    "    exclude_names = [\"Enterobacter cloacae complex sp.\", \"Pseudomonas fluorescens group sp.\", \"Bacillus cereus group sp.\", \"Citrobacter freundii complex sp.\"]\n",
    "    for name in exclude_names:\n",
    "        if organism.startswith(name):\n",
    "            return False\n",
    "    org_split = organism.split()\n",
    "    if len(org_split)==1:\n",
    "        return False\n",
    "    else:\n",
    "        first_word, second_word = org_split[0:2]\n",
    "    if second_word in [\"sp.\", \"bacterium\", \"cf.\", \"endosymbiont\", \"symbiont\", \"genomosp.\"]:   # Clostridiales bacterium,  Cyanosarcina cf.    Spiroplasma endosymbiont     Type-E symbiont of Plautia stali     Agrobacterium genomosp. 3 str. CFBP 6623\n",
    "        return False\n",
    "    first_letter = first_word[0]\n",
    "    if first_letter.islower():\n",
    "        return False\n",
    "    if first_word in [\"Candidatus\", \"uncultured\", \"SAR324\"]:  # synthetic, unidentified prokaryote, SAR324 cluster bacterium\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "011479c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing uncultured and unidentified prokaryotes\n",
    "df_uniq_genome=df_uniq_genome[df_uniq_genome.apply(is_valid_name, axis = 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7bec39ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing duplicates from species_taxid to have only one from each species_taxid\n",
    "df_uniq_genome = df_uniq_genome.sample(frac=1).drop_duplicates(subset=\"species_taxid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7d176883",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5686, 39)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniq_genome.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8ba67e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniq_genome.to_csv(\"genebank_test_data/genebank_prok_1_per_species.tsv\", header=True, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "81d19650",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniq_genome.to_csv(\"genebank_test_data/genebank_prok_1_per_species_accession.tsv\", columns=[\"assembly_accession\"], header=False, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b4a45157",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please adjust the [#$ -t] in run_dfastqc_GeneBank.sh to [#$ -t 5686]\n"
     ]
    }
   ],
   "source": [
    "number_of_genomes = df_uniq_genome.shape[0]\n",
    "print(rf\"Please adjust the [#$ -t] in run_dfastqc_GeneBank.sh to [#$ -t {number_of_genomes}]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d7fef603",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dummy data for testing\n",
    "df_uniq_genome_dummy = df_uniq_genome.head(5)\n",
    "df_uniq_genome_dummy.to_csv(\"genebank_test_data/genebank_prok_1_per_species_dummy.tsv\", columns=[\"assembly_accession\"],  header=True, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f3afbd61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dummy data for testing\n",
    "df_uniq_genome_dummy = df_uniq_genome.head(5)\n",
    "df_uniq_genome_dummy.to_csv(\"genebank_test_data/genebank_prok_1_per_species_dummy.tsv\", columns=[\"assembly_accession\"],  header=True, index=False, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
